src/service/xhr-impl.js

```diff
@@ -189,8 +203,39 @@ export class Xhr {
       init.headers['Content-Type'] = 'application/json;charset=utf-8';
       init.body = JSON.stringify(init.body);
     }
-    return this.fetch(input, init)
-        .then(response => response.json());
+
+    const isCacheable = (init.method === 'GET' && propertyPath);
+    if (isCacheable) {
+      const cacheKey = this.getCacheKey_(input, opt_init);
+      const cachedPromise = this.fragmentCache_.get(cacheKey);
+      if (cachedPromise) {
+        return cachedPromise.then(getPropertyAtPath);
+      }
+
+      const fetchPromise = this.fetch(input, init).then(getResponseJson);
+
+      // Since a fragment is present, cache the full response promise, then
+      // return a promise with just the value specified by the fragment.
+      this.fragmentCache_.put(cacheKey, fetchPromise);
+      return fetchPromise.then(getPropertyAtPath);
+    } else {
+      return this.fetch(input, init).then(getResponseJson);
+    }
+  }
+
+  /**
+   * Creates a cache key for a fetch.
+   *
+   * @param {string} url
+   * @param {?FetchInitDef=} opt_init
+   * @return {string}
+   * @private
+   */
+  getCacheKey_(url, opt_init) {
+    const urlWithoutHash = url.slice(0, url.indexOf('#'));
```

Comment:
Danger! `url.indexOf` can return `-1`, which'll make this be something like:

```js
const url = 'https://example.com';
const urlWithoutHash = url.slice(0, url.indexOf('#')); // => https://example.co
```

We have a [`removeFragment`](https://github.com/ampproject/amphtml/blob/master/src/url.js#L277-L289) helper to do this.